[EPIC 1.1] Access Control v2 - Google Auth Sign In/Up

liamohkay opened the pull request:
## Related Issue
#193 Access Control V2 - Implement Google Auth, captcha, cookies and error handling

## Description

### Breaking Changes:
1. Google authentication must be enabled on production firebase.
2. A production [Google Recaptcha Key](https://www.google.com/recaptcha/admin/create) must be created.
Environment variables `NEXT_PUBLIC_RECAPTCHA_SITE_KEY` and `RECAPTCHA_SECRET` must be filled in in order for the reCAPTCHA api requests to succeed. These variables have been added in the `sample.env` with `FILL_ME_IN` values.

### Main Features/Additions:
- Created `@/templates/SignInForm/GoogleSignInButton.jsx` 
Button signs in or signs up user via Firebase Google Auth. A user will be automatically signed up if it does not exist, therefore the same logic & component can be used for both sign in and sign up cases. This button was added to `@/templates/SignInForm/SignInForm.jsx` and `@/templates/SignUpForm/SignUpForm.jsx`

- Created `handleGoogleLogin` method in `@/libs/providers/GlobalProvider.jsx`
Google + Firebase auth logic. Updates global auth state & user redux store. Handles onboarding rerouting logic. Used as clickHandler in `GoogleSignInButton.jsx`. Google Auth Provider is given to the entire app in `_app` children.

- Created `@/templates/utils/OrBlock.jsx`
Renders a line break with 'Or' text surrounded by a line
Used on SignIn & SignUp forms in between authentication methods.

- Created `verifyRecaptcha` method
Method uses Google reCAPTCHA v3 method to invisibly verify a human user using Google's API. Allows for a callback function to be executed after verifcation from google API is successful. Throws error and does not execute callback function if verification is unsuccessful.
Added to all sign in & sign up methods.
Unfortunately cannot be abstracted out to a re-useable  function as the callback function may be undefined if dlecared out of scope of certain global providers.

- Created `@/templates/pages/api/verify-recaptcha.js` API route
API post route that is called in `verifyRecaptcha`, sends token to Google API & returns a success / fail based on if user is human or not.

All methods use the same global auth provider and redux store operations correctly update authentication state (same auth provider as previous sign in/up methods).

All methods have proper `try/catch` error handling and logging and the API route handles multiple error cases with correct res codes.

### More Info
- Session cookies has been pushed until Next13 upgrade per discussion with @Ahmedr275 as firebase already handled user session cookies.
`@/libs/utils/cookieUtils.js` has been pushed to be used with session cookies when Next 13 upgrade happens. None of the methods in `cookieUtils.js` are utilized in this PR, and are here merely for future use.

-  **Testing**
I have unit tests for the sign in/up button and verifyRecaptcha method, but am waiting confirmation on what the jest config should be before uploading. Happy to upload them when a answer is given on the config.

## Type of Change

Please select the type(s) of change that apply and delete those that do not. 

- [x] **New feature**: A non-breaking change that adds functionality.
- [x] **Breaking change**: A change that causes existing functionality to not work as expected.
- [x] **Documentation update**: Changes or updates to documentation.

## Proposed Solution

Describe your code changes in detail. Explain how you implemented your solution and any design decisions you made. 

## How to Test

 Provide instructions on how to test these changes. Include details on test configurations, test cases, and expected outcomes. 

## Unit Tests

 List the unit tests added or modified to verify your changes. 

## Documentation Updates

- [x] Yes **1/2/25 documentation has been updated.**

Documentation needs to be added on how to sign up for [Google reCAPTCHA](https://www.google.com/recaptcha/admin/create). The free tier gets 10,000 free requests per month.

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

Acceptance Criteria
- [x] Google Authentication is enabled and working for both sign-in and sign-up.
- [x] Users can log in and sign up using their Google accounts seamlessly.
- [x] Captcha verification is integrated and functional on all login and sign-up attempts.
- [ ] Session cookies comply with GDPR and other privacy standards.
- [x] User-friendly error messages are displayed for all authentication issues.
- [x] Errors are logged appropriately for monitoring and debugging.
- [x] Thorough testing is completed, and all edge cases are handled effectively.

liamohkay committed:
[EPIC 1.1] Added google sign in/up, added reCPATCHA v3 verifcation & api route, added global provider & redux store helpers

Cosmodocus commented on templates/SignIn/SignInForm/GoogleSignInButton.jsx:
It seems like this buttons functionality is more or less not just to handle signing in users with google but signing them up as well. Maybe a more specific naming convention for the button would make more sense in this case, such as 'GoogleButton' or 'GoogleAuthButton'. You can ask chatgpt if necessary lol.

And if that is also the case regarding the functionality, it would be best to consider a different placement for this button, such as within the components folder in the root directory.

liamohkay replied:
Changed to `GoogleAuthButton`

Cosmodocus commented on pages/api/verify-recaptcha.js:
So this logic seems sound so far, great job. However this doesn't really follow the stucturing of our platform. Please remove the api folder from pages, and move this logic somewhere where it's best suited for, such as under libs/services/users/ where we house other third party api functions

liamohkay replied:
@Cosmodocus Ok I am pretty new to NextJS, my only question mark here with this change is would moving it out of an API call and moving it expose the CAPTCHA secret key on the front end in compiled code?

Cosmodocus commented on templates/utils/OrBlock.jsx:
I'm not quite sure what's going on here. This seems to be a UI component that is placed in the utils folder created within the templates folder. If you notice other folders/entry files in the templates folder, this doesn't really follow a similar pattern, and if it only serves to be a reusable component, would be best placed into our components folder.

A utils file typically contains helper functions that are primarily logic-based, with no direct connection to the UI

liamohkay replied:
Moved to `components/OrBlock/OrBlock.jsx`

Cosmodocus commented on templates/SignIn/SignInForm/GoogleSignInButton.test.jsx:
We discussed the possibility of reworking the unit testing approach and flow in the future, and I understand you're aware that the changes made here may need to be reverted to align with those updates. 

Once the unit test feature is ready, we could also follow that convention by merging it into this branch, and following its patterns & structures.

liamohkay replied:
Im going to remove the tests, jest config & associated library installs. I think a top down config for tests from the maintainer team is the best forward that way every PR doesnt have a different config.

Cosmodocus reviewed (changes requested):
Hey @liamohkay, I left some comments that you can take a look at. Feel free to ask any questions. 

It looks good so far but there are some things that require changes.

Also, I would greatly appreciate a video of any sort of you going through the feature, as well as a testable link in which I can access and manually test the feature myself.

Cosmodocus commented on templates/SignIn/SignInForm/GoogleSignInButton.jsx:
```diff
@@ -0,0 +1,46 @@
+
+import { AuthContext } from '@/libs/providers/GlobalProvider';
+
+const GoogleSignInButton = ({ isSignIn }) => {
```
Lets use JSDocs to comment the components functionality, as well as the type & description of the props used. Refer to other components if you need assistance on this.

liamohkay replied:
JS docs added for google button & the recaptcha function.

Cosmodocus commented on templates/SignIn/SignInForm/GoogleSignInButton.jsx:
```diff
@@ -0,0 +1,46 @@
+
+import { AuthContext } from '@/libs/providers/GlobalProvider';
+
+const GoogleSignInButton = ({ isSignIn }) => {
```
We like to destructure the props of components into a props variable. This will keep this component more readable, maintainable, and scalable. These things will matter as this platform continues to grow 

Just do the following:
`const GoogleSignInButton = ( props ) => {`
`const { isSignIn } = props`

liamohkay replied:
Fixed.

Cosmodocus commented on templates/utils/OrBlock.jsx:
```diff
@@ -0,0 +1,35 @@
+import { Grid } from '@mui/material';
+import { styled } from '@mui/system';
+
```
You can add a JSDocs comment for this as well

liamohkay replied:
Added JSdocs.

Cosmodocus commented on libs/utils/CookieUtils.js:
Looks good as far as I can tell. Lets leave some JSDocs comments for each utility function here.

liamohkay committed:
Implemented PR fixes, removed testing

liamohkay commented:
> Hey @liamohkay, I left some comments that you can take a look at. Feel free to ask any questions.
> 
> It looks good so far but there are some things that require changes.
> 
> Also, I would greatly appreciate a video of any sort of you going through the feature, as well as a testable link in which I can access and manually test the feature myself.

Hey @Cosmodocus . I implemented the changes you requested. I also removed the tests, jest configs, and associated packages. The only change that was not implemented was moving the `verify-recaptcha` from the `/api` folder. It needs to b e an api route for a backend FN, otherwise if its imported like the other third party API functions the secret key will be compiled on the front end and expose it.

liamohkay committed:
Implemented PR fixes, removed testing
